Sdx.Web.Url の修正用PR

sd-tomita opened the pull request:
## 概要
#16 の修正用PRです。
## TODO

※元PRからコピペ。何かあれば随時追記
- [x] Get、Set メソッド廃止してプロパティを使うようにする
- [x] Sdx.Web.Url.~~Protocol~~ Scheme を追加する(列挙型を使う)
- [x] 文字列連結の数が多い箇所で、foreach を使わずに StringBuilder を使う
- [x] foreach → ForEach / LINQ に書き換え(LINQはできるかどうかの確認も含め)
- [x] クエリなしURLを渡した際にエラーになるのを修正する

sd-tomita committed:
BuildQueryString() 内で引数の型で処理を分岐していたのをやめる

以前の動作に戻しています。
BuildQueryString() は、あくまで「引数を元にURL文字列を生成する」動作だけを
担当することになるので、メソッドの責任としてもこのほうがよさそうです。

sd-tomita committed:
Dictionary の連結処理を LINQ を使ったコードに変更

sd-tomita committed:
URL文字列の組み立てを StringBuilder() を使った方法に変更

String クラスは文字列操作のたびに新しいオブジェクトが生成されるらしい。
StringBuilder クラスの場合はオブジェクトの内容を変更するだけで
新しいオブジェクトは作らないようなので、こちらに変更しました。

sd-tomita committed:
無駄な Dictionary を作っている箇所があったので削除

ついでにコメントも何か変だったので修正

sd-tomita committed:
this.paraData に何もオブジェクト入ってなかったので追加

完全に凡ミス

sd-tomita committed:
foreach ステートメントを List.ForEach で書き換えました

中の処理で特に戻り値が無いので今回は Select 等の出番は無さそうと思い
普通に ForEach を使いました。

sd-tomita committed:
引数が String[] 時の処理を追加

List用の処理から単に引数の型を変えるだけでいけた。

gomo commented on Sdx.Web/Url.cs:
```diff
@@ -108,21 +77,29 @@ private string BuildPath()
+      ;
+      string query = this.BuildQueryString(excludedDic);
       return path + query;
     }
```
おお！この辺のディクショナリを加工する処理綺麗ですね。ザC#って感じ :+1: 
C#をばっちり書き慣れてる人のコードって感じがします。

細かいですけど、この２つの引数がそれぞれ`addDic`とか`excludeList`という名前なのは冗長な感じがいます。引数の型宣言があるので、`add`/`exclude`で十分な感じがします。

普段なら治す程でもないけど、Sdxだし、公開インターフェースなので、直しておいてください。

:innocent: 

また

``` c#
var addedDic = this.paramData
       .Concat(addDic)
       .ToDictionary(p => p.Key, p => p.Value)
```

というのが何をやってるか一瞬分かりづらいかったです。というのも`addedDic`と`addDic`というのが見分けがつきませんでした。一時変数なので単純にresultsとparams(これダメなんだっけ？queryParamsとか？)とするか、あるいは

``` c#
      string query = this.BuildQueryString(
         this.paramData
          .Concat(addDic)
          .ToDictionary(p => p.Key, p => p.Value)
      );
```

っていうのも有りかも。こちらは、ローカル変数なんで直さなくてもOKです。

gomo replied:
Concatで重複したキーが合った時どうなるのか動作確認してください。重複した場合上書きになるよう実装お願いします。

:innocent:

sd-tomita replied:
わかりやすく例外になりましたｗ
上書きになるように修正しますー

```
テスト名:   TestMethod1
テスト成果:    失敗
結果  のメッセージ: 
テスト メソッド UnitTest.UrlTest.TestMethod1 が例外をスローしました:
System.ArgumentException: 同一のキーを含む項目が既に追加されています。
結果  の標準出力:    TestMethod1
```

sd-tomita committed:
Concat でキーが重複した場合の挙動を確認するテストを追加

結果例外だったので修正が必要そうです。

sd-tomita committed:
Concat でエラーになるため、事前に重複キーを持つパラメータを除外

sd-tomita committed:
変数の整理

・冗長な変数名を修正
・不要な変数を削除

sd-tomita committed:
必要そうなプロパティを追加

とりあえず Visual Studio 先生の機能で生成だけしてみました。
エラーになるのでテストは一旦全てコメントアウトしています。

sd-tomita committed:
簡単なプロパティを追加

機能を使ってみる目的で、簡単そうな Host, Path の2つを
まずは追加してみました

sd-tomita committed:
パラメータ用のプロパティを追加

sd-tomita committed:
不要な Get,Set メソッドを削除

プロパティに移行したため削除しました

sd-tomita committed:
通信プロトコル用データ追加

sd-tomita committed:
Scheme の取得テストを追加

sd-tomita committed:
Scheme の取得方法を変更

Protocol だと意味が異なってしまうので
Scheme で取得できるように変更しました

sd-tomita commented:
`"http://example.com/path/to/api/?hoge=fuga&foo=bar"` というURL文字列から通信プロトコルの取得って無理なんじゃないだろうか？あくまで文字列しか情報ないし。。。こういうのはなんとなく Request クラスあたりから取り出すイメージがある。要確認。

gomo commented:
> URL文字列から通信プロトコルの取得って無理なんじゃないだろうか？あくまで文字列しか情報ないし。。。

恐らく何か勘違いしてますね。URLでプロトコルを表してる部分は`http`あるいは`https`という部分ですのでそこを見ればわかります。列挙しがそうなんってるんだから、察してくれｗ

正確にはスキームといいます。「WEBのURLのスキームは通信プロトコルを表している」という表現が正確。そういう意味では、列挙しやメソッド名もSchemeの方がいいかな。そうすれば、WEB以外のURLにも対応させられるし。

:innocent:

matsukawa1 commented:
@tomita-ychr 

.NET Framework 4.6 and 4.5からは、下記のようなメソッドがあるみたいですね。
HttpWorkerRequest.GetProtocol メソッド
https://msdn.microsoft.com/ja-jp/library/system.web.httpworkerrequest.getprotocol(v=vs.110).aspx

---

javaだと、下記記事を見る限り、getProtocol()とgetScheme()の2種類のメソッドがあるみたいですね。
http://www.javadrive.jp/servlet/request/index7.html

`mailto:user@dammy.com`の場合は、
getScheme() → mailto(スキーム)
getProtocol() → SMTP や ESMTP(プロトコル)
が取得できるのでしょうね(おそらく)。

なので、`URL文字列から通信プロトコルの取得って無理`ってのは考え方としては間違ってはいないと思います。(getProtocol() は、スキームからプロトコルを判別しているだけだったりしてw)

ただ、通信プロトコルを「判別」するだけなら、スキームで事足りるのだと思います。

> 「WEBのURLのスキームは通信プロトコルを表している」という表現が正確。

↑ようはこれが全てな感じですねw

tomita-ychrさんのコメントが個人的に気になったので、少し調べてみましたw

gomo commented:
> なので、URL文字列から通信プロトコルの取得って無理ってのは考え方としては間違ってはいないと思います。(getProtocol() は、スキームからプロトコルを判別しているだけだったりしてw)
> 
> ただ、通信プロトコルを「判別」するだけなら、スキームで事足りるのだと思います。

Wiki曰く

> スキーム名としてはプロトコル名が用いられていることが多いがそれに限らない。

https://ja.wikipedia.org/wiki/Uniform_Resource_Locator

ということですね。なので最初に私が「プロトコル」と言ってしまったのが正確じゃなかったですね。

ただ、

`URL文字列から通信プロトコルの取得って無理`は正しいが、`"http://example.com/path/to/api/?hoge=fuga&foo=bar" というURL文字列から通信プロトコルの取得って無理`は間違いなので、

> 恐らく何か勘違いしてますね。

と思ったわけです。

sd-tomita commented:
@gomo @matsukawa1 いろいろお騒がせしてしまってすみませんでした。

> 列挙しやメソッド名もSchemeの方がいいかな。そうすれば、WEB以外のURLにも対応させられるし。

こちら了解です。

sd-tomita committed:
テスト内容の変更

"Protocol" → "Scheme" に変更するにあたってテスト内容を変更。
なお、列挙体とプロパティの名前がかぶってしまうので、プロパティについては
名前を Name と Type に分けました。

Sdx.Web.Url の中に作らず、Sdx.Web.Scheme にすればかぶらないですが
Url クラス内の Scheme にしないと、http,https 以外のスキームもないと
意味がおかしくなるので、とりあえずこのようにしています。

sd-tomita committed:
とりあえずテストが通るようにURLスキーム周りの処理を実装

Url.SchemeType は Sdx.Web.Url.Scheme 自体が返ってくるような
イメージがするので名前が微妙かも。
Sdx.Web.Url.Scheme.Http --> "http" という取得ができたほうが
使い勝手は良さそう。ただそれならDictionaryで済むような気がするので
かえって微妙かもしれない。

ひとまずテストが通ったので一旦コミット

sd-tomita commented on Sdx.Web/Url.cs:
```diff
@@ -10,120 +11,138 @@ public class Url
     private Uri uri;
     private Dictionary<string, string> paramData;
 
+    public enum Scheme
```
`System.Uri.Scheme` プロパティからスキームは取得できるので、列挙体使う方法はとりあえずやめます。

sd-tomita committed:
スキーム管理方法変更

列挙子は使わないことになったのでコードを修正

sd-tomita committed:
System.Uri はコンストラクター以外では使わないようにしました

各プロパティが read-only で、もしあとからスキームだけを変更したりしたい場合に
対応できなかったので、メンバ変数に保存しておくようにしました。

sd-tomita committed:
不要な using の削除

sd-tomita committed:
クエリの無いURLをコンストラクタの引数に渡した際のテストを追加

後から順次セットしていく使い方もありうるのでテスト

sd-tomita committed:
テストをもう1パターン追加

こういうケースもあるかもしれない

sd-tomita committed:
パラメータが単数、なし の場合にエラーになっていたのを修正

gomo commented on Sdx.Web/Url.cs:
```diff
@@ -7,123 +7,161 @@ namespace Sdx.Web
+      {
+        this.scheme = value;
+      }
     }
```
``` c#
    public string Scheme
    {
      get
      {
        return this.scheme;
      }

      set
      {
        this.scheme = value;
      }
    }
```

このプロパティーは

``` c#
    public string Scheme{ get; set; }
```

とかけます。尚且つ、メンバー変数を作る必要もありません。

このコードでも動くので間違いじゃないけど、SDX内だし、覚えるという意味でも:innocent:

gomo commented on UnitTest/UrlTest.cs:
```diff
@@ -42,17 +42,47 @@ public void TestMethod1()
+      //各部品の取得
+      Assert.Equal("example.com", url.Domain);
+      Assert.Equal("http", url.Scheme);
+      Assert.Equal("/path/to/api", url.LocalPath);
```
パラメーターがNULLなのかあるけど空なのか、テストを入れたほうがいいですね。テストは文字通りテストの役目もあるのですが、生きたドキュメントの役目も果たすので、そういう観点でテストを書くといいです。

:innocent:

gomo replied:
あと、
- カラの状態から、パラメータを追加する
- カラの状態から、Buildにパラメータを渡す
- カラの状態から、Excludeパラメータを渡す

のテストもあったほうがいいです。3番めはどうするのがいいのか考えてみてください。

:innocent:

gomo commented on UnitTest/UrlTest.cs:
```diff
@@ -42,17 +42,47 @@ public void TestMethod1()
+    public void TestNonParams()
     {
-      Console.WriteLine("TestMethod2");
+      Console.WriteLine("TestNonParams");
```
テスト内にConsoleを残しておくのは、今のところ問題無さそうなので私もやってますが、ふと気になったのは、テストのConsoleはメソッドごと別々にみれなかったっけ？だとすると余り意味が無いと思いました。

gomo commented on UnitTest/UrlTest.cs:
```diff
@@ -42,17 +42,47 @@ public void TestMethod1()
+      //各部品の取得
+      Assert.Equal("example.com", url.Domain);
+      Assert.Equal("http", url.Scheme);
+      Assert.Equal("/path/to/api", url.LocalPath);
```
こちらも、上の3つと同じようになテストを足してください。

:innocent: 

他にしておいた方がいいテスト無いかね？テストはしっかりしてあった方がいいに決まってますが、最悪テストが漏れていてバグが有った時もその機能をつかって開発中に気づくことが多いので、その時に直せばいいけど。

基本的にこのようなライブラリのコードの修正は下記のように実行します。
1. テストに漏れがあってそこにバグが有り、クライアントコードでそのバグが発覚。
2. cs-sdxに戻ってきてまず、そのバグが発生するテストを書く。
3. そのテストが通るように実装する。
4. 他のテストを実行して、今回の修正で他の部分が壊れていないかチェックする。

まず、見つかったバグのテストを書くことが重要です。テスト駆動開発ですね。

gomo commented:
ケアレスミスだと思うけど最初のテストの名前がTestMethod1のままですね。

他のと合わせるとすると`TestMultipleParams`とかですかね？

:innocent:

sd-tomita commented:
差分がわからなくなるので、このブランチは proj にマージして、
いただいたコメントへの修正は別ブランチで改めて行うようにします。
